### PR TITLE
Add SNIP 52

### PR DESCRIPTION
**Pull request by blake-regalia**
This PR adds SNIP-52: Private Push Notifications.

SNIP-52 supports two operating modes, one of which is called "TxHash" mode. The upcoming v1.10 release of Secret Network includes a feature that makes the transaction hash available to the env in order to support this mode 🎉

Reference impl (contract): https://github.com/SolarRepublic/snip52-reference-impl
Example client: https://github.com/SolarRepublic/snip52-demo
Demo: https://demo.starshell.net/

### CODE REVIEW

**Comment by blake-regalia** [comment]
@assafmo @Cashmaney

File: SNIP-52.md [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+
+  // produce the nonce by XOR'ing the two previous 12-byte results
+  let nonce := xorBytes(channelIdBytes, counterBytes)
+
+  // construct the additional authenticated data
+  let aad := concatStrings(env.blockHeight, ":", env.txHash)
+
+  // split payload
+  let ciphertext := slice(payload, len(payload) - 16)
+  let tag := slice(payload, len(ciphertext))
```
Comment (eshelB) [inline SNIP-52.md:570]: here I think there's an error as it seems that the `tag` and the `ciphertext` are receiving the same part of the payload, the tag should be the second element of the slice

File: SNIP-52.md [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+      "method": "subscribe",
+      "params": {
+        "query": "wasm.contract_address='secret1ku936rhzqynw6w2gqp4e5hsdtnvwajj6r7mc5z'"
+      }
+    }
+    ```
+
+    Alice will now receive a JSONRPC message for each new execution of the contract, from which she can search for her unique Notification ID.
+    
+    <!-- Alternatively, if she trusts the WebSocket server won't record her activity, she can use a filter in the `query` field of her subscription message, e.g., `wasm.ZjZjYzVhYjU4 EXISTS` . -->
```
Comment (eshelB) [inline SNIP-52.md:109]: Why is this commented out? It does not appear in the markdown rendering of the document

**Review by toml01 — COMMENTED** [review]
Overall looks great!
Left a few notes/thoughts

File: SNIP-52.md (hunk 1 above)
Comment (blake-regalia) [reply SNIP-52.md:570]: good catch 👌

File: SNIP-52.md [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+```
+
+If the channel is operating in TxHash Mode, given by `"mode": "txhash"`, then the response includes the Notification ID of the next event in the given channel affecting the given viewer (who is specified in the authentication data, depending on whether a query permit or viewer key is used).
+
+The response also provides the viewer's current seed for the given channel, allowing the client to derive future Notification IDs for this channel offline (i.e., without having to query the contract again).
+
+
+## UpdateSeed Method
+
+Allows clients to set a new shared secret. In order to guarantee the provided secret has high entropy, clients must submit a signed document params and signature to be verified before the new shared secret (a SHA-256 hash of the signature) is accepted.
```
[the remark this answers is not in the document]
Comment (blake-regalia) [reply SNIP-52.md:307]: This is a fair point to bring up. These were the motivations for choosing the permit-like signature approach:

1. it prevents a malicious contract developer from "back door-ing" their way into a user's seed, say, if the contract developer got ahold of, or was able to manipulate, the internal contract secret.
2. it enforces high entropy. if we allowed the client to provide its own entropy, a poor choice such as "password123" would put _other_ users (e.g., the sender of a notification) at risk of privacy leak as well.
3. it naturally creates a sequence of deterministic seeds, meaning that a user could in theory recover previous seeds in order to decrypt older tx events.

File: SNIP-52.md (hunk 2 above)
Comment (blake-regalia) [reply SNIP-52.md]: Thanks, it was put on hold during the transition to Counter vs TxHash mode. It has been updated.

File: SNIP-52.md [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+
+Each channel can operate in one of two modes: Counter Mode, or TxHash Mode.
+
+There is a trade-off between these two modes. Counter Mode is easier for clients but less secure.
+
+In Counter Mode:
+ - ✅ clients only have to recompute a channel's notification ID each time they receive a notification from that channel
+ - ✅ clients are able to use node APIs such as `tx_search` to search back through history and find a missed notification
+ - ✅ clients are able to query the contract to obtain their next notification ID, allowing them to bypass much of the SNIP-52 client-side implementation
+ - ❌ an attacker could hypothetically de-anonymize notification IDs using a sophisticated side-chain attack
```
[the remark this answers is not in the document]
Comment (blake-regalia) [reply SNIP-52.md]: Added a link to the replay attacks section on docs.scrt.network, which should provide enough helpful context to curious readers I think. The actual attack is not described in-depth anywhere.

File: SNIP-52.md [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+
+Request:
+```json
+{
+  "update_seed": {
+    "signed_doc": {
+      "params": {
+        "chain_id": "secret-4",
+        "contract": "<bech32 address of contract>",
+        "previous_seed": "<base64-encoded value of previous seed>"
```
[the remark this answers is not in the document]
Comment (blake-regalia) [reply SNIP-52.md]: You're right. Thanks for pointing this out

**Pull request merged**